### PR TITLE
Scopes: Adapt for new API

### PR DESCRIPTION
**Pull request by bfmatei**
**What is this feature?**

The Scopes API moved to a node-based approach. This PR brings compatibility for the UI to use that client.

**Why do we need this feature?**

Implement Scopes?

**Who is this feature for?**

-

**Which issue(s) does this PR fix?**:

-

**Special notes for your reviewer:**

- Video:

https://github.com/grafana/grafana/assets/9215315/2d52fe43-0e67-4543-bf88-2f39aff62288

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: packages/grafana-prometheus/src/datasource.ts [hunk 1]
```diff
@@ -598,6 +598,11 @@ export class PrometheusDatasource
   // it is used in metric_find_query.ts
   // and in Tempo here grafana/public/app/plugins/datasource/tempo/QueryEditor/ServiceGraphSection.tsx
   async getTagKeys(options: DataSourceGetTagKeysOptions<PromQuery>): Promise<MetricFindValue[]> {
+    return Promise.resolve([
+      { value: 'key', text: 'Key' },
+      { value: 'anotherKey', text: 'anotherKey' },
+    ]);
```
Comment (dprokop) [inline packages/grafana-prometheus/src/datasource.ts:604]: Please remove this before merging ;)

File: packages/grafana-prometheus/src/datasource.ts [hunk 2]
```diff
@@ -620,6 +625,11 @@ export class PrometheusDatasource
 
   // By implementing getTagKeys and getTagValues we add ad-hoc filters functionality
   async getTagValues(options: DataSourceGetTagValuesOptions<PromQuery>) {
+    return Promise.resolve([
+      { value: 'value', text: 'Value' },
+      { value: 'anotherValue', text: 'anotherValue' },
+    ]);
```
Comment (dprokop) [inline packages/grafana-prometheus/src/datasource.ts:631]: Please remove this before merging ;)

Btw - for testing you can use static dimensions, variables can be configured with those directly in the UI.

File: public/app/features/apiserver/server.ts [hunk 3]
```diff
@@ -69,7 +75,7 @@ export class ScopedResourceServer<T = object, K = string> implements ResourceSer
 
           case 'in':
           case 'notin':
-            return `${key}${operator}(${label.value.join(',')})`;
+            return `${key} ${operator} (${label.value.join(',')})`;
```
Comment (dprokop) [inline public/app/features/apiserver/server.ts:78]: Is this change required? I see a similar line for = / != operators above that was not changed.

File: public/app/features/dashboard-scene/scene/ScopesFiltersScene.tsx [hunk 4]
```diff
@@ -9,111 +10,269 @@ import {
...
+                  evt.stopPropagation();
+
+                  if (linkId) {
+                    model.toggleScope(linkId);
+                  }
+                }}
+              />
+            )}
+
+            <span>{node.item.title}</span>
```
Comment (dprokop) [inline public/app/features/dashboard-scene/scene/ScopesFiltersScene.tsx:242]: UX: I miss the possibility to click on node item, I need to be very precise to click on the checkbox.

File: public/app/features/dashboard-scene/scene/ScopesFiltersScene.tsx [hunk 5]
```diff
@@ -9,111 +10,269 @@ import {
...
+            tabIndex={0}
+            className={cx(styles.item, isScope && styles.itemScope)}
+            onClick={(evt) => {
+              evt.stopPropagation();
+              model.expandNode(nodePath);
+            }}
+            onKeyDown={(evt) => {
+              evt.stopPropagation();
+              model.expandNode(nodePath);
+            }}
```
Comment (dprokop) [inline public/app/features/dashboard-scene/scene/ScopesFiltersScene.tsx:224]: Think we need to be more thoughtful here, currently the API is being hit very eagerly. For example, I have my focus (after tabbing into it) on the first node (Applications):

image

I would expect I can navigate using the keyboard, but instead I'm stuck on the Application node and pressing arrow down just re-executes the same API (`http://localhost:3000/apis/scope.grafana.app/v0alpha1/namespaces/default/find?parent=applications`)

File: public/app/features/dashboard-scene/scene/ScopesScene.tsx [hunk 6]
```diff
@@ -27,11 +27,11 @@ export class ScopesScene extends SceneObjectBase<ScopesSceneState> {
...
+      this.state.filters.fetchBaseNodes();
 
       const filtersValueSubscription = this.state.filters.subscribeToState((newState, prevState) => {
-        if (newState.value !== prevState.value) {
-          this.state.dashboards.fetchDashboards(newState.value);
+        if (newState.scopes !== prevState.scopes) {
+          this.state.dashboards.fetchDashboards(newState.scopes);
           sceneGraph.getTimeRange(this.parent!).onRefresh();
         }
       });
```
Comment (dprokop) [inline public/app/features/dashboard-scene/scene/ScopesScene.tsx:37]: Is the long term plan to have the UX improved for the initial scopes loading? Currently with scopes preserved in the URL I cannot reason about them when opening the scopes selector for the first time, as we only fetch top leel scopes:

image


Would be great if we could figure out which sub trees to open based on the persisted scopes.

**Review by dprokop — COMMENTED** [review]
@bfmatei - the overall flow works good I think, and API adjustments are fine.

However,  I have concerns about the UX, on top of the comments that I've left:

1. Any change in the scope selector (selecting a scope) results in new queries. I think we may want to defer that until the scopes selector is closed.
2. Would love the entire scope selector row to be clickable rather than the checkbox only.
3. I miss some placeholder on the empty scope selector, currently it's just an empty input.


Would be also good if we added some tests for the tree navigation UI to make sure we are not breaking anything in the future.

File: public/app/features/dashboard-scene/scene/ScopesScene.tsx (hunk 6 above)
Comment (bfmatei) [reply public/app/features/dashboard-scene/scene/ScopesScene.tsx:37]: We had a discussion around this yesterday (get the entire path for a given scope name) but that's something that requires a design doc and more thinking on how to do it. For the time being we agreed on being OK with the current mechanism.

File: public/app/features/dashboard-scene/scene/ScopesFiltersScene.tsx (hunk 4 above)
Comment (bfmatei) [reply public/app/features/dashboard-scene/scene/ScopesFiltersScene.tsx:254]: Yeah, I tried to work around this even for the showcase but it gets a little weird from an accessibility perspective. Once we have the final designs, I'll find a way to overcome this while implementing them.

File: public/app/features/dashboard-scene/scene/ScopesScene.tsx (hunk 6 above)
Comment (dprokop) [reply public/app/features/dashboard-scene/scene/ScopesScene.tsx:37]: great!

File: public/app/features/dashboard-scene/scene/ScopesFiltersScene.tsx (hunk 5 above)
Comment (dprokop) [reply public/app/features/dashboard-scene/scene/ScopesFiltersScene.tsx]: yeah, understand that!

**Review by dprokop — APPROVED** [review]
